### PR TITLE
array fix

### PR DESCRIPTION
**Pull request by memoryleak47**

### CODE REVIEW

File: lang/step.md [hunk 1]
```diff
@@ -202,15 +202,15 @@ impl<M: Memory> Machine<M> {
...
+                if index >= 0 && index < count {
                     index * elem.size::<M>()
                 } else {
                     throw_ub!("out-of-bounds array access");
                 }
             }
             _ => panic!("index projection on non-indexable type"),
         };
-        assert!(offset < ty.size::<M>());
+        assert!(offset < ty.size::<M>() || ty.size::<M>().is_zero());
```
Comment (memoryleak47) [inline lang/step.md:213]: This failed for ZSTs.

File: lang/step.md [hunk 2]
```diff
@@ -202,15 +202,15 @@ impl<M: Memory> Machine<M> {
         };
         let offset = match ty {
             Type::Array { elem, count } => {
-                if index < count {
+                if index >= 0 && index < count {
```
Comment (memoryleak47) [inline lang/step.md:205]: Negative indices were not considered out-of-bounds before.

File: lang/step.md [hunk 3]
```diff
@@ -190,7 +190,7 @@ impl<M: Memory> Machine<M> {
             Type::Union { fields, .. } => fields[field].0,
             _ => panic!("field projection on non-projectable type"),
         };
-        assert!(offset < ty.size::<M>());
+        assert!(offset < ty.size::<M>() || ty.size::<M>().is_zero());
```
Comment (memoryleak47) [inline lang/step.md:193]: I think this will still fail when accessing the second field of `(i8, ())`.
Maybe get rid of this assert altogether?

**Comment by RalfJung** [comment]
Those assertions should probably be offset<=size.

**Comment by memoryleak47** [comment]
@RalfJung ready

**Comment by RalfJung** [comment]
Nice finds! Make sure to record the testcases that found these issue, we'll want them on CI.

**Pull request merged**